docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md

```diff
@@ -0,0 +1,202 @@
+
+### Data model and wire protocol
+
+Prometheus is a dimensional time-series database. Each time-series is represented by a series of Samples (timestamp and value), identified by a set of unique Labels (key, value). Labels represents the dimensions with one label named `__name__` playing the special role of metric name. The set of Labels represent the unique identifier of the time-series. Records (Samples) are simply the values (doubles) in order of timestamp (long, as UNIX timestamp since the Epoch).
+
+<center>
+<br/>
+<img src="/img/blog/2024-11-26-introducing-new-prometheus-connector/prometheus-data-model.png" width="60%"/>
+<br/>
+Fig. 1 - Logical data model of the Prometheus TSDB.
+</center>
+
+The wire-protocol payload defined by the [Remote-Write specifications](https://prometheus.io/docs/specs/remote_write_spec/) allows batching in a single write request datapoints (samples) from multiple time-series, and multiple samples for each time-series. This the Protobuf definition of the single method exposed by the interface:
+
+```
+func Send(WriteRequest)
+
+message WriteRequest {
+  repeated TimeSeries timeseries = 1;
+  // [...]
+}
+
+message TimeSeries {
+  repeated Label labels   = 1;
+  repeated Sample samples = 2;
+}
+
+message Label {
+  string name  = 1;
+  string value = 2;
+}
+
+message Sample {
+  double value    = 1;
+  int64 timestamp = 2;
+}
+```
+
+The POST payload is serialized as Protobuf 3, compressed with [Snappy](https://github.com/google/snappy).
+
+### Specifications constraints
+
+Specifications do not impose any constraints on repeating `TimeSeries` with the same set of `Label`, but it does require strict ordering of `Sample` belonging to the same time-series (by `timestamp`) and `Label` in a single `TimeSeries` (by name).
+
+{{< hint info >}}
+The term "time-series" is overloaded, referring to both:
+1. A unique series of samples in the datastore, identified by unique set of labels,
+    and
+2. `TimeSeries` as a block of the `WriteRequest`.
+
+The two concepts are obviously related, but a `WriteRequest` may contain multiple `TimeSeries` elements referring to the same datastore *time-series*.
+{{< /hint >}}
+
+### Prometheus design principles vs Flink design principles
+
+Flink is designed with a consistency-first approach. By default, any unexpected error causes the streaming job to fail and restart from the last checkpoint.
+
+In contrast, Prometheus is designed with an availability-first approach, prioritizing fast ingestion over strict consistency. When a write request contains malformed entries, the entire request must be discarded and not retried. Additionally, samples belonging to the same time series (with the same dimensions or labels) must be written in strict timestamp order.
+
+You may have already spotted the issue: any malformed or out-of-order samples can act as a “poison pill” unless you drop the offending request and proceed.
```

Comment:
How does this differ from a malformed write to any other sink? Is it because AsyncSink does not guarantee ordering?

Reply:
Prometheus treats malformed and out-of-order the same way: 400 (rejected).
This sink has fixed in-flight request = 1 to prevent causing out-of-order while retrying